File: src/main/java/ru/ezhov/rocket/action/infrastructure/ReflectionRocketActionUiRepository.java
package ru.ezhov.rocket.action.infrastructure;

import org.reflections.Reflections;
import org.reflections.scanners.SubTypesScanner;
import ru.ezhov.rocket.action.RocketActionUiRepository;
import ru.ezhov.rocket.action.api.RocketActionUi;

import java.lang.reflect.Modifier;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Set;

public class ReflectionRocketActionUiRepository implements RocketActionUiRepository {
    private List<RocketActionUi> list;

    @Override
    public List<RocketActionUi> all() {
        if (list == null) {
            list = new ArrayList<>();

            Reflections reflections = new Reflections("", new SubTypesScanner(true));
            final Set<Class<? extends RocketActionUi>> classes = reflections.getSubTypesOf(RocketActionUi.class);
            for (Class aClass : classes) {
                try {
                    if (!Modifier.isAbstract(aClass.getModifiers())) {
                        list.add((RocketActionUi) aClass.newInstance());
                    }
                } catch (InstantiationException e) {
                    e.printStackTrace();
                } catch (IllegalAccessException e) {
                    e.printStackTrace();
                }
            }
        }

        return list;
    }

    @Override
    public Optional<RocketActionUi> by(String type) {
        final List<RocketActionUi> all = all();
        return all.stream().filter(r -> r.type().equals(type)).findFirst();
    }
}

File: src/main/kotlin/ru/ezhov/rocket/action/infrastructure/ReflectionRocketActionUiRepository.kt
package ru.ezhov.rocket.action.infrastructure

import mu.KotlinLogging
import org.reflections.Reflections
import org.reflections.scanners.Scanners
import org.reflections.util.ClasspathHelper
import org.reflections.util.ConfigurationBuilder
import ru.ezhov.rocket.action.api.RocketActionFactoryUi
import ru.ezhov.rocket.action.api.RocketActionType
import ru.ezhov.rocket.action.domain.RocketActionUiRepository
import java.lang.reflect.Modifier

private val logger = KotlinLogging.logger {}

class ReflectionRocketActionUiRepository : RocketActionUiRepository {
    private var list: MutableList<RocketActionFactoryUi> = mutableListOf()
    fun load() {
        val reflections = Reflections(
                ConfigurationBuilder()
                        .addUrls(ClasspathHelper.forPackage("ru.ezhov.rocket.action"))
                        .setScanners(Scanners.SubTypes)
        )
        val classes = reflections.getSubTypesOf(RocketActionFactoryUi::class.java)
        for (aClass in classes) {
            try {
                if (!Modifier.isAbstract(aClass.modifiers)) {
                    list.add(aClass.getConstructor().newInstance() as RocketActionFactoryUi)
                }
            } catch (e: InstantiationException) {
                logger.error(e) { "Error when load ui" }
            } catch (e: IllegalAccessException) {
                logger.error(e) { "Error when load ui" }
            }
        }
    }

    override fun all(): List<RocketActionFactoryUi> = list

    override fun by(type: RocketActionType): RocketActionFactoryUi? =
            all().firstOrNull { r: RocketActionFactoryUi -> r.type().value() == type.value() }
}